bxserum/provider/grpc.go

```diff
@@ -63,10 +72,72 @@ func NewGRPCClientWithOpts(opts RPCOpts) (*GRPCClient, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &GRPCClient{
-		apiClient:  pb.NewApiClient(conn),
-		privateKey: opts.PrivateKey,
-	}, nil
+	client := &GRPCClient{
+		apiClient:                     pb.NewApiClient(conn),
+		privateKey:                    opts.PrivateKey,
+		recentBlockHashExpiryDuration: opts.CacheBlockHash,
+	}
+	go func() {
+		ctx := context.Background()
+		stream, err := client.GetRecentBlockHashStream(ctx)
+		if err != nil {
+			log.Error("can't open recent block hash stream")
+			return
+		}
+		ch := stream.Channel(1)
+		for {
+			select {
+			case hash := <-ch:
+				client.updateBlockHash(hash)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return client, nil
+}
+
+func (g *GRPCClient) updateBlockHash(hash *pb.GetRecentBlockHashResponse) {
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+	g.recentBlockHash = hash
+	now := time.Now()
+	g.recentBlockHashTime = &now
+}
+
+func (g *GRPCClient) RecentBlockHash(ctx context.Context) (*pb.GetRecentBlockHashResponse, error) {
+	response := g.getCachedBlockHash()
+	if response != nil {
+		return response, nil
+	}
+
+	g.mutex.Lock()
+	defer g.mutex.Unlock()
+
+	now := time.Now()
+	if g.recentBlockHash != nil && g.recentBlockHashTime.Before(now.Add(g.recentBlockHashExpiryDuration)) {
```

Comment:
think the `.Before` call is backward –– should be `g.recentBlockHashTime.Add(g.recentblockHashExpiryDuration).Before(now)`?

Reply:
`g.recentBlockHashTime` is already always before `now`